ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/PostgresLedgerDao.scala

```diff
@@ -718,26 +718,36 @@ private class PostgresLedgerDao(
     ~ ledgerString("transaction_id")
     ~ ledgerString("workflow_id").?
     ~ date("effective_at")
+    ~ list[String]("witnesses").map(_.map(Party.assertFromString))
+    ~ list[List[String]]("divulgences").map(_.map {
+      case List(party, tid) =>
+        Party.assertFromString(party) -> TransactionIdString.assertFromString(tid)
+    }.toMap)
+    ~ list[String]("maintainers").map(_.map(Party.assertFromString))
     ~ binaryStream("contract")
     ~ binaryStream("key").? map (flatten))
 
   private val SQL_SELECT_CONTRACT =
     SQL(
-      "select c.*, le.effective_at from contracts c inner join ledger_entries le on c.transaction_id = le.transaction_id where id={contract_id} and archive_offset is null ")
-
-  private val SQL_SELECT_WITNESS =
-    SQL("select witness from contract_witnesses where contract_id={contract_id}")
-
-  private val DivulgenceParser = (party("party")
-    ~ long("ledger_offset")
-    ~ ledgerString("transaction_id") map (flatten))
-
-  private val SQL_SELECT_DIVULGENCE =
-    SQL(
-      "select party, ledger_offset, transaction_id from contract_divulgences where contract_id={contract_id}")
-
-  private val SQL_SELECT_KEY_MAINTAINERS =
-    SQL("select maintainer from contract_key_maintainers where contract_id={contract_id}")
+      """
+        |select
+        |   c.*,
+        |   le.effective_at,
+        |   array_agg(witness) witnesses,
+        |   coalesce(array_agg(array[d.party, d.transaction_id]) filter (where d.party is not null), array[]::varchar[][]) divulgences,
+        |   case when c.key is not null
+        |        then array(select m.maintainer from contract_key_maintainers m where m.contract_id = c.id)
+        |	       else array[]::varchar[]
+        |   end maintainers
+        |from contracts c
+        |   inner join ledger_entries le
+        |       on c.transaction_id = le.transaction_id
```

Comment:
Just a heads up: We should depend on as little as possible on the ledger entries table and store all required data in the "ACS" tables to be prepared to support participant index pruning (time-based trimming of the ledger entries table).

Reply:
Another thing we should soon fix is being able to insert divulged contracts into the index, which obviously won't have associated transaction (we wouldn't even know the transaction id). Would probably be good to get together and discuss how the database schema needs to evolve to work as the index database for a participant with partial view on the ledger...